### DISCUSSION TITLE
How to download latest private GitHub release with PowerShell script?

### DISCUSSION THREAD

**Discussion by bmils** [community/community#24683] [Apps, API and Webhooks]
<p>Hello,</p>
<p>I recently wrote a PowerShell script that downloads the latest release from a <strong>public</strong> repo and that works as intended. However, I want to change my script so it can access my <strong>private</strong> repo. Here is the code I have tried so far:</p>
<pre><code class="lang-auto"># Download latest release from GitHub
$credentials="myPersonalAccessToken"
$headers = New-Object "System.Collections.Generic.Dictionary[[String],[String]]"
$headers.Add("Authorization", "token $credentials")

$repo = "myUserName/MyPrivateReleaseRepo"
$file = "MyBinaries.zip"

$releases = "https://api.github.com/repos/$repo/releases"

Write-Host Determining latest release
$tag = (Invoke-WebRequest $releases -Headers $headers | ConvertFrom-Json)[0].tag_name

$download = "https://github.com/$repo/releases/download/$tag/$file"
$name = $file.Split(".")[0]
$zip = "$name-$tag.zip"
$dir = "$name-$tag"

Write-Host Dowloading latest release
Invoke-WebRequest $download -Out $zip

Write-Host Extracting release files
Expand-Archive $zip -Force

# Cleaning up target dir
Remove-Item "C:\MyOutPutFolder\$name" -Recurse -Force -ErrorAction SilentlyContinue 

# Moving from temp dir to target dir
Move-Item $dir\$name -Destination "C:\MyOutPutFolder\$name" -Force

# Removing temp files
Remove-Item $zip -Force
Remove-Item $dir -Recurse -Force
</code></pre>
<p>I get the following error when testing on my private repo:</p>
<pre><code class="lang-auto">Invoke-WebRequest : Not Found
At C:\Script\DownloadLatestGitHubRelease.ps1:25 char:1
+ Invoke-WebRequest $download -Out $zip
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
</code></pre>
<p>Am I using the access token incorrectly? As I said above the script works perfectly fine on my public repo and fails on my private repo.</p>

**Comment by sonibla** [comment, marked as answer]
<p>Hello there <a class="mention" href="/bmils">@bmils</a></p>
<p>I found a solution !</p>
<p>Simple version (without tag and folder): <a href="https://gist.github.com/sonibla/a60fc31b244ceba3220b9bb33316798c" class="inline-onebox" rel="noopener nofollow ugc">Download assets from private Github releases using PowerShell · GitHub</a></p>
<p>Version similar to your script:</p>
<pre><code class="lang-auto"># Download latest release from GitHub
$credentials="&lt;github_access_token&gt;"
$repo = "&lt;user_or_org&gt;/&lt;repo_name&gt;"
$file = "&lt;name_of_asset_file&gt;"
$folder="&lt;folder_to_extract_to&gt;"

$headers = New-Object "System.Collections.Generic.Dictionary[[String],[String]]"
$headers.Add("Authorization", "token $credentials")
$releases = "https://api.github.com/repos/$repo/releases"

Write-Host Determining latest release
$id = ((Invoke-WebRequest $releases -Headers $headers | ConvertFrom-Json)[0].assets | where { $_.name -eq $file })[0].id
$tag = (Invoke-WebRequest $releases -Headers $headers | ConvertFrom-Json)[0].tag_name

$download = "https://" + $credentials + ":@api.github.com/repos/$repo/releases/assets/$id"
$name = $file.Split(".")[0]
$zip = "$name-$tag.zip"
$dir = "$name-$tag"

Write-Host Dowloading latest release
$headers.Add("Accept", "application/octet-stream")
$zip = "$name-$tag.zip"
Invoke-WebRequest -Uri $download -Headers $headers -OutFile $zip

Write-Host Extracting release files
Expand-Archive $zip -Force

# Cleaning up target dir
Remove-Item "$folder\$name" -Recurse -Force -ErrorAction SilentlyContinue 

# Moving from temp dir to target dir
Move-Item $dir\$name -Destination "$folder\$name" -Force

# Removing temp files
Remove-Item $zip -Force
Remove-Item $dir -Recurse -Force
</code></pre>
<p>The Linux version helped me a lot: <a href="https://gist.github.com/maxim/6e15aa45ba010ab030c4" class="inline-onebox" rel="noopener nofollow ugc">Download assets from private Github releases · GitHub</a></p>

**Comment by livingod101** [comment]
Is this still working for you? It does not work for me, get error:

Invoke-WebRequest : {"message":"Not Found","documentation_url":"https://docs.github.com/rest/releases/releases#list-releases"}
At line:11 char:9

**Reply by sonibla** [reply]
Yes it's still working (I actually use this script in a CI that runs multiple times a day)
To debug you could print the value of $id and check it actually match a file to download
Other troubleshooting ideas:
Maybe the token you're using doesn't give access to the repository/releases of the repository?
Or the latest release doesn't have any asset?
